[Compiled Autograd] Fix duplicate visits of same node

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106234
* #106176
* __->__ #105887

The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

jansel committed:
[Compiled Autograd] Fix duplicate visits of same node

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel commented:
N

> I'm not sure why `before()` is called multiple times? Isn't it only called during a particular Node's apply?
> 
> Is this due to the fake "Add" Node that you add to the graph not behaving properly (I don't recall the details on how that works).

I explained it in the PR description:  It happens when we have multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.

`test_duplicate_backward_root` is a good example of this:
https://github.com/pytorch/pytorch/blob/9e4e0ecdd9cd78c8bdb0aeaa35c33ecd6ab9e832/test/test_autograd.py#L2526

We need to visit the nodes pointed to by edges for handle their input metadata for type promotion and sum_to_size.

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

albanD commented on torch/csrc/dynamo/python_compiled_autograd.cpp:
```diff
@@ -423,6 +423,7 @@ variable_list compiled_autograd(
             return ss.str();
           });
       saved.after(call.node->next_edges());
+      saved.debug_asserts();
```
Can we add this above between the apply and the before on the next_edges to make sure the apply didn't leave the saved in a bad state?

jansel replied:
done

albanD commented on torch/csrc/dynamo/compiled_autograd.h:
```diff
@@ -470,54 +470,44 @@ struct TraceState {
 
   void before(c10::SymInt& t) {
-    SWAP_SAVED_VARIABLES_SAVE(stashed_symints, t, t);
+    stashed_symints.save(&t, c10::SymInt(t));
```
Why are we copying the SymInt here?

jansel replied:
`save()` requires either `std::move()` or copy due to the `&&` arg.  We need the std::move() in the other cases because SavedVariable doesn't support copy construction, but don't want it here.

albanD reviewed (commented):
Ho ok, you do use before/after outside of the apply function. Thanks for the details.

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

jansel committed:
Update on "[Compiled Autograd] Fix duplicate visits of same node"


The error fixed here happened when we had multiple autograd::Edge objects pointing to the same autograd::Node, causing before() to get called multiple times on the same object.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

jansel commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge
